### PR TITLE
Resplit package to avoid cyclic dependency

### LINKED ISSUE #26: libwebp-base should add a run_constrained on libwebp version
Otherwise, the following environment is possible,

```
  libwebp            conda-forge/osx-64::libwebp-1.0.2-hd3bf737_5
  libwebp-base       conda-forge/osx-64::libwebp-base-1.1.0-2
```

### PR DESCRIPTION
**Pull request by carterbox**
@isuruf I resplit the package using [curl as an example](https://github.com/AnacondaRecipes/curl-feedstock/blob/master/recipe/meta.yaml). The result is that Windows now has two packages (although one is empty), and the base package has `run_constrained` to hopefully solve #26.

@ocefpaf I moved `libtiff` out of the `requirements/host` section in order to isolate it from the install PREFIX. This allows `libwebp` to be built against libtiff even though there is a circular dependency which would cause PREFIX to already have the previous build of `libwebp` installed.

Checklist
* [x] Used a [fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

Resolves #26

### CODE REVIEW

**Comment by carterbox** [comment]
I realized that C/C++ does not have native support for optional dependencies at runtime, so I'm going to add all the runtime dependencies back. However, we still have a problem with the circular dependency between `libtiff` and `libwebp`. Perhaps build variants is the answer?

**Comment by ocefpaf** [comment]
@isuruf does this one looks good to you?

**Comment by isuruf** [comment]
@ocefpaf, can you run `conda render recipe -m .ci_support/linux_.yaml` and post the output?

**Comment by isuruf** [comment]
@ocefpaf, can you check now?

**Comment by isuruf** [comment]
I mean post the `conda render` output.

**Comment by ocefpaf** [comment]
> I mean post the `conda render` output.

It fails with:

```
> conda render recipe -m .ci_support/linux_.yaml                                                                                                                                            ⏎
Traceback (most recent call last):
  File "/home/filipe/miniconda3/bin/conda-render", line 11, in <module>
    sys.exit(main())
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/conda_build/cli/main_render.py", line 228, in main
    return execute(sys.argv[1:])
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/conda_build/cli/main_render.py", line 198, in execute
    variants=args.variants)
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/conda_build/api.py", line 52, in render
    bypass_env_check=bypass_env_check):
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/conda_build/metadata.py", line 2054, in get_output_metadata_set
    bypass_env_check=bypass_env_check)
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/conda_build/metadata.py", line 727, in finalize_outputs_pass
    permit_unsatisfiable_variants=permit_unsatisfiable_variants)
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/conda_build/render.py", line 535, in finalize_metadata
    m = parent_metadata.get_output_metadata(m.get_rendered_output(m.name()))
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/conda_build/metadata.py", line 2112, in get_rendered_output
    permit_undefined_jinja=permit_undefined_jinja, variant=variant):
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/conda_build/metadata.py", line 2101, in get_rendered_outputs_section
    variant=variant)) or
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/conda_build/metadata.py", line 1534, in _get_contents
    rendered = template.render(environment=env)
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/jinja2/environment.py", line 1090, in render
    self.environment.handle_exception()
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "<template>", line 14, in top-level template code
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/conda_build/jinja_context.py", line 259, in pin_compatible
    "Is it one of your host dependencies?".format(package_name))
RuntimeError: Could not get compatibility information for jpeg package.  Is it one of your host dependencies?
```

**Comment by ocefpaf** [comment]
OK. With your last commit it works:

```
> conda render recipe -m .ci_support/linux_.yaml 
--------------
Hash contents:
--------------
{'c_compiler': 'gcc', 'c_compiler_version': '7'}
----------
meta.yaml:
----------
package:
    name: libwebp-base
    version: 1.1.0
source:
    sha256: 98a052268cc4d5ece27f76572a7f50293f439c17a98e67c4ea0c7ed6f50ef043
    url: http://storage.googleapis.com/downloads.webmproject.org/releases/webp/libwebp-1.1.0.tar.gz
build:
    number: 3
    run_exports:
        - libwebp-base >=1.1.0,<1.2.0a0
requirements:
    build:
        - libstdcxx-ng 9.2.0 hdf63c60_2
        - binutils_impl_linux-64 2.34 h53a641e_0
        - ld_impl_linux-64 2.34 h53a641e_0
        - _openmp_mutex 4.5 0_gnu
        - gcc_impl_linux-64 7.3.0 hd420e75_5
        - libgomp 9.2.0 h24d8f2e_2
        - libgcc-ng 9.2.0 h24d8f2e_2
        - _libgcc_mutex 0.1 conda_forge
        - gcc_linux-64 7.3.0 h553295d_18
        - binutils_linux-64 2.34 hc952b39_18
    host:
        - _libgcc_mutex 0.1 conda_forge
        - libgomp 9.2.0 h24d8f2e_2
        - _openmp_mutex 4.5 0_gnu
        - libgcc-ng 9.2.0 h24d8f2e_2
    run:
        - libgcc-ng >=7.3.0
    run_constrained:
        - libwebp >=1.1,<1.2.0a0
test:
    commands:
        - test -f $PREFIX/lib/libwebp.a
        - test -f $PREFIX/lib/libwebp.so
        - test -f $PREFIX/include/webp/decode.h
        - test -f $PREFIX/include/webp/encode.h
        - test -f $PREFIX/include/webp/types.h
about:
    description: 'WebP is a method of lossy and lossless compression that can be used
        on a large variety of photographic,

        translucent and graphical images found on the web. The degree of lossy compression
        is adjustable so a

        user can choose the trade-off between file size and image quality.

        '
    dev_url: https://chromium.googlesource.com/webm/libwebp
    doc_url: https://developers.google.com/speed/webp/docs/using
    home: https://developers.google.com/speed/webp
    license: BSD-3-Clause
    license_file: COPYING
    summary: WebP image library
extra:
    copy_test_source_files: true
    final: true
    parent_recipe:
        name: webp_split_recipe
        path: /home/filipe/00-CONDA-FORGE/bot/libwebp-feedstock/recipe
        version: 1.1.0
    recipe-maintainers:
        - dschreij

--------------
Hash contents:
--------------
{'c_compiler': 'gcc',
 'c_compiler_version': '7',
 'giflib': '5.2',
 'jpeg': '9',
 'libpng': '1.6',
 'libtiff': '4.1.0'}
----------
meta.yaml:
----------
package:
    name: libwebp
    version: 1.1.0
source:
    sha256: 98a052268cc4d5ece27f76572a7f50293f439c17a98e67c4ea0c7ed6f50ef043
    url: http://storage.googleapis.com/downloads.webmproject.org/releases/webp/libwebp-1.1.0.tar.gz
build:
    number: 3
    run_exports:
        - libwebp-base >=1.1.0,<1.2.0a0
requirements:
    build:
        - libstdcxx-ng 9.2.0 hdf63c60_2
        - binutils_impl_linux-64 2.34 h53a641e_0
        - ld_impl_linux-64 2.34 h53a641e_0
        - _openmp_mutex 4.5 0_gnu
        - gcc_impl_linux-64 7.3.0 hd420e75_5
        - libgomp 9.2.0 h24d8f2e_2
        - libgcc-ng 9.2.0 h24d8f2e_2
        - _libgcc_mutex 0.1 conda_forge
        - gcc_linux-64 7.3.0 h553295d_18
        - binutils_linux-64 2.34 hc952b39_18
    host:
        - _libgcc_mutex 0.1 conda_forge
        - libgomp 9.2.0 h24d8f2e_2
        - _openmp_mutex 4.5 0_gnu
        - libgcc-ng 9.2.0 h24d8f2e_2
    run:
        - jpeg 9.*
        - giflib 5.2.*
        - libwebp-base >=1.1.0,<1.2.0a0
        - libgcc-ng >=7.3.0
        - libpng 1.6.*
        - libtiff 4.1.0.*
test:
    commands:
        - test -f $PREFIX/bin/cwebp
        - test -f $PREFIX/bin/dwebp
        - dwebp examples/test.webp
    source_files:
        - examples/test.webp
about:
    description: 'WebP is a method of lossy and lossless compression that can be used
        on a large variety of photographic,

        translucent and graphical images found on the web. The degree of lossy compression
        is adjustable so a

        user can choose the trade-off between file size and image quality.

        '
    dev_url: https://chromium.googlesource.com/webm/libwebp
    doc_url: https://developers.google.com/speed/webp/docs/using
    home: https://developers.google.com/speed/webp
    license: BSD-3-Clause
    license_file: COPYING
    summary: WebP image library
extra:
    copy_test_source_files: true
    final: true
    parent_recipe:
        name: webp_split_recipe
        path: /home/filipe/00-CONDA-FORGE/bot/libwebp-feedstock/recipe
        version: 1.1.0
    recipe-maintainers:
        - dschreij
```

**Comment by isuruf** [comment]
Can you post again? This should work now

**Comment by ocefpaf** [comment]
> Can you post again? This should work now

```
> conda render recipe -m .ci_support/linux_.yaml 
Traceback (most recent call last):
  File "/home/filipe/miniconda3/bin/conda-render", line 11, in <module>
    sys.exit(main())
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/conda_build/cli/main_render.py", line 228, in main
    return execute(sys.argv[1:])
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/conda_build/cli/main_render.py", line 198, in execute
    variants=args.variants)
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/conda_build/api.py", line 52, in render
    bypass_env_check=bypass_env_check):
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/conda_build/metadata.py", line 2057, in get_output_metadata_set
    ensure_matching_hashes(conda_packages)
  File "/home/filipe/miniconda3/lib/python3.7/site-packages/conda_build/metadata.py", line 333, in ensure_matching_hashes
    "Involved packages were:\n" + error)
conda_build.exceptions.RecipeError: Mismatching hashes in recipe. Exact pins in dependencies that contribute to the hash often cause this. Can you change one or more exact pins to version bound constraints?
Involved packages were:
Mismatching package: libwebp-base (id h516909a_3); dep: libwebp-base 1.1.0 2; consumer package: libwebp
```

**Comment by isuruf** [comment]
I think the only way to fix this is to have 2 separate feedstocks for `libwebp` and `libwebp-base`.

**Comment by carterbox** [comment]
@isuruf The tests were passing with my last commit. What are you trying to accomplish with your changes?

**Comment by isuruf** [comment]
Because your run dependencies are wrong.
```
requirements:
    build:
        - _libgcc_mutex 0.1 conda_forge
        - gcc_linux-64 7.3.0 h553295d_18
        - binutils_linux-64 2.34 hc952b39_18
        - libgcc-ng 9.2.0 h24d8f2e_2
        - binutils_impl_linux-64 2.34 h53a641e_0
        - _openmp_mutex 4.5 0_gnu
        - libstdcxx-ng 9.2.0 hdf63c60_2
        - ld_impl_linux-64 2.34 h53a641e_0
        - gcc_impl_linux-64 7.3.0 hd420e75_5
        - libgomp 9.2.0 h24d8f2e_2
    host:
        - _libgcc_mutex 0.1 conda_forge
        - libgomp 9.2.0 h24d8f2e_2
        - _openmp_mutex 4.5 0_gnu
        - libgcc-ng 9.2.0 h24d8f2e_2
    run:
        - libpng
        - giflib
        - jpeg
        - libtiff
        - libgcc-ng >=7.3.0
        - libwebp-base >=1.1.0,<1.2.0a0
```

**Comment by isuruf** [comment]
@ocefpaf, @carterbox, can you send a PR to staged-recipes to add `libwebp-base` feedstock?

**Comment by carterbox** [comment]
The problem is the circular dependency between libtiff and libwebp because it breaks the jinja pinnings template. Creating another libwebp feedstock does not solve this problem.

libtiff must be modified to remove the runtime requirement on libwebp. Instead it should depend on libwebp-base.

Broken: webp ➡️ tiff ➡️ webp
Fixed: webp ➡️ tiff ➡️ webp-base

**Comment by isuruf** [comment]
`libwebp-base` feedstock will not depend on `libtiff`. Then `libtiff` can be built only with `libwebp-base`. Finally `libwebp` will depend on both `libwebp-base` and `libtiff`. There will be no cyclic dependencies then.

**Comment by carterbox** [comment]
The libweb-base output in this recipe *already* doesn't depend on libtiff.

> Finally libwebp will depend on both libwebp-base and libtiff.

And libtiff *still* [depends on libwebp](https://github.com/conda-forge/libtiff-feedstock/blob/4f5172ec6a1f8b8114d8d80593e4ad2e68d628c2/recipe/meta.yaml#L59)! Which means the circle is not broken!

**Comment by isuruf** [comment]
> And libtiff still depends on libwebp! 

No. `libtiff` will depend on `libwebp-base`.

**Comment by isuruf** [comment]
While I appreciate your effort in this PR, it's still broken because of the cyclic dependencies.

**Comment by carterbox** [comment]
Great. We are agreed! libtiff needs to be changed so it only depends on libwebp-base! :smiley: After that adding pinnings to this PR should work.

**Comment by carterbox** [comment]
@isuruf, You are looking for these dependencies? I'm not sure which libraries you want pinned to the minor revision.
```
----------
meta.yaml:
----------
package:
    name: libwebp
    version: 1.1.0
source:
    sha256: 98a052268cc4d5ece27f76572a7f50293f439c17a98e67c4ea0c7ed6f50ef043
    url: http://storage.googleapis.com/downloads.webmproject.org/releases/webp/libwebp-1.1.0.tar.gz
build:
    number: 3
    run_exports:
        - libwebp-base >=1.1.0,<1.2.0a0
requirements:
    build:
        - binutils_linux-64 2.34 hc952b39_18
        - binutils_impl_linux-64 2.34 h53a641e_0
        - _openmp_mutex 4.5 0_gnu
        - _libgcc_mutex 0.1 conda_forge
        - ld_impl_linux-64 2.34 h53a641e_0
        - libstdcxx-ng 9.2.0 hdf63c60_2
        - gcc_linux-64 7.3.0 h553295d_18
        - libgcc-ng 9.2.0 h24d8f2e_2
        - gcc_impl_linux-64 7.3.0 hd420e75_5
        - libgomp 9.2.0 h24d8f2e_2
    host:
        - _libgcc_mutex 0.1 conda_forge
        - libstdcxx-ng 9.2.0 hdf63c60_2
        - libgomp 9.2.0 h24d8f2e_2
        - _openmp_mutex 4.5 0_gnu
        - libgcc-ng 9.2.0 h24d8f2e_2
        - giflib 5.2.1 h516909a_2
        - jpeg 9c h14c3975_1001
        - lz4-c 1.8.3 he1b5a44_1001
        - xz 5.2.4 h14c3975_1001
        - zlib 1.2.11 h516909a_1006
        - libpng 1.6.37 hed695b0_1
        - zstd 1.4.4 h3b9ef0a_2
        - libtiff 4.0.10 hc3755c2_1005
    run:
        - libgcc-ng >=7.3.0
        - libwebp-base >=1.1.0,<1.2.0a0
        - jpeg >=9c,<10a
        - libpng >=1.6.37,<1.7.0a0
        - giflib >=5.2.1,<5.3.0a0
        - libtiff >=4.0.10,<5.0a0
```

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,91 +1,97 @@
...
+# section defines the build environment. The outputs/requirements are only for
+# determining the exported runtime constraints for each subpackage.
 requirements:
   build:
     - {{ compiler('c') }}
     - make  # [not win]
-  host:
-    - jpeg  # [not win]
-    - libpng  # [not win]
     - libtiff  # [not win]
```
Comment (isuruf) [inline recipe/meta.yaml]: This should be in `host`.

**Comment by isuruf** [comment]
This feedstock as a whole still has a cyclic dependency with libtiff feedstock.
For example, when a new libwebp major version comes out, we can't build this feedstock because `libtiff` depends on the previous major version of `libwebp`. So, my suggestion is to have a new feedstock `libwebp-base`.

**Comment by carterbox** [comment]
That's a very good reason to have a separate feedstock. I guess conda environments don't let you have multiple versions of the same library installed.

I believe this feedstock would successfully build, but you would have to run it a second time after libtiff was rebuilt so that libwebp was built against a libtiff that was built against a libwebp-base of the same major version. Anyways, this current recipe could be used as a temporary measure until another feedstock is created.

**Comment by isuruf** [comment]
> I believe this feedstock would successfully build, but you would have to run it a second time after libtiff was rebuilt so that libwebp was built against a libtiff that was built against a libwebp-base of the same major version.

How can `libtiff` be built against a new major version of `libwebp-base` when the new major of version of `libwebp-base` cannot be built?

File: recipe/meta.yaml (hunk 1 above)
Comment (carterbox) [reply recipe/meta.yaml]: I left a very detailed commit message about why I moved libtiff to the build section.
cc592bd7985437809dbb997a0429b264149e9cb7

Comment (isuruf) [reply recipe/meta.yaml]: Another reason to have a separate feedstock for `libwebp-base`

**Comment by isuruf** [comment]
I get the following,
```
--------------
Hash contents:
--------------
{'c_compiler': 'gcc', 'c_compiler_version': '7'}
----------
meta.yaml:
----------
package:
    name: libwebp
    version: 1.1.0
source:
    sha256: 98a052268cc4d5ece27f76572a7f50293f439c17a98e67c4ea0c7ed6f50ef043
    url: http://storage.googleapis.com/downloads.webmproject.org/releases/webp/libwebp-1.1.0.tar.gz
build:
    number: 3
    run_exports:
        - libwebp-base >=1.1.0,<1.2.0a0
requirements:
    build:
        - ld_impl_linux-64 2.34 h53a641e_0
        - libgomp 9.2.0 h24d8f2e_2
        - binutils_impl_linux-64 2.34 h53a641e_0
        - gcc_linux-64 7.3.0 h553295d_18
        - libgcc-ng 9.2.0 h24d8f2e_2
        - gcc_impl_linux-64 7.3.0 hd420e75_5
        - binutils_linux-64 2.34 hc952b39_18
        - _libgcc_mutex 0.1 conda_forge
        - libstdcxx-ng 9.2.0 hdf63c60_2
        - _openmp_mutex 4.5 0_gnu
    host:
        - _libgcc_mutex 0.1 conda_forge
        - libgomp 9.2.0 h24d8f2e_2
        - _openmp_mutex 4.5 0_gnu
        - libgcc-ng 9.2.0 h24d8f2e_2
    run:
        - libgcc-ng >=7.3.0
        - libtiff
        - giflib
        - jpeg
        - libwebp-base >=1.1.0,<1.2.0a0
        - libpng
```

**Comment by carterbox** [comment]
You're right; a new feedstock is necessary before the next major version. The linking would be like `libwebp=2 -> libtiff=1 -> libwebp-base=1`, so you wouldn't be able to install it and then the tests would fail which prevents `libwebp-base=2` from being published.

**Comment by carterbox** [comment]
IDK, I rerendered my local feedstock, and now I'm getting this error:

```
conda_build.exceptions.RecipeError: Mismatching hashes in recipe. Exact pins in dependencies that contribute to the hash often cause this. Can you change one or more exact pins to version bound constraints?
Involved packages were:
Mismatching package: libwebp-base (id h516909a_3); dep: libwebp-base 1.1.0 h14c3975_3; consumer package: libwebp
```

I'm going to give up because it's actually super late for me. I may decide I have time to create a `libwebp-base-feedstock` tomorrow. Thanks for troubleshooting this with me.

:moon: :sleeping_bed:

**Comment by isuruf** [comment]
Thanks @carterbox for the PR and the discussion.

**Comment by ocefpaf** [comment]
> I think the only way to fix this is to have 2 separate feedstocks for `libwebp` and `libwebp-base`.

Sad that outputs cannot be used here :-(

@carterbox are you sending the PR to staged-recipes? If so please ping me there.

**Comment by carterbox** [comment]
I volunteer to send the PR to staged recipes sometime before Sunday.

**Comment by ocefpaf** [comment]
Thanks @carterbox! I'm swamped with work and would not be able to do so before that.

**Pull request closed without merging**